### PR TITLE
Allow testing kdump on SLE Micro

### PR DESCRIPTION
**Pull request by Martchus**
* Based on #17805 which makes the kdump setup work on transactional systems
    * I was not able to verify #17805 due to https://bugzilla.suse.com/show_bug.cgi?id=1213721 so this PR is still a draft. However, with the additional commit to allow testing kdump on SLE Micro as well I was able to verify that it works on SLE Micro.
* See particular commit messages and https://progress.opensuse.org/issues/136028

### CODE REVIEW

**Comment by Martchus** [comment]
grafik
grafik
grafik

File: lib/kdump_utils.pm [hunk 1]
```diff
@@ -321,10 +363,14 @@ sub configure_service {
...
 
     prepare_for_kdump(%args);
-    if ($args{yast_interface} eq 'cli') {
-        activate_kdump_cli;
+    if (is_transactional) {
+        activate_kdump_transactional;
     } else {
-        return 16 if (activate_kdump == 16);
+        if ($args{yast_interface} eq 'cli') {
+            activate_kdump_cli;
```
Comment (czerw) [inline lib/kdump_utils.pm:370]: I would rather do decision here:
`is_transactional ? activate_kdump_transactional : activate_kdump_cli;`

File: lib/kdump_utils.pm [hunk 2]
```diff
@@ -385,9 +431,15 @@ sub check_function {
     if ($args{test_type} eq 'function') {
         # Check, that vmcore exists, otherwise fail
         assert_script_run('ls -lah /var/crash/*/vmcore', 240);
-        my $vmlinux = (is_sle("<16") || is_leap("<16.0")) ? '/boot/vmlinux-$(uname -r)*' : '/usr/lib/modules/$(uname -r)/vmlinux*';
-        my $crash_cmd = "echo exit | crash `ls -1t /var/crash/*/vmcore | head -n1` $vmlinux";
-        validate_script_output "$crash_cmd", sub { m/PANIC:\s([^\s]+)/ }, is_aarch64 ? 1200 : 800;
+        if (is_sle_micro || is_leap_micro) {
```
Comment (czerw) [inline lib/kdump_utils.pm:434]: wouldn't be `is_transactional` better?

File: lib/kdump_utils.pm [hunk 3]
```diff
@@ -34,8 +33,23 @@ sub install_kernel_debuginfo {
     # kernel-default(x86-64) = 5.3.18-lp152.26.2
     # kernel-default-base(x86-64) = 5.3.18-lp152.26.2.lp152.8.2.2
     # -> kernel-default-debuginfo-5.3.18-lp152.26.2
-    my $debuginfo = script_output('rpm -qf /boot/initrd-$(uname -r) --provides | awk \'match($0,/(kernel-.+)\(.+\) = (.+)/,m) {printf "%d %s-debuginfo-%s\n", length($0), m[1], m[2]}\' | sort -n | head -n1 | cut -d" " -f2-');
-    zypper_call("-v in $debuginfo", timeout => 4000);
+    return script_output('rpm -qf /boot/initrd-$(uname -r) --provides | awk \'match($0,/(kernel-.+)\(.+\) = (.+)/,m) {printf "%d %s-debuginfo-%s\n", length($0), m[1], m[2]}\' | sort -n | head -n1 | cut -d" " -f2-');
+}
+
+sub install_kernel_debuginfo {
```
Comment (czerw) [inline lib/kdump_utils.pm:39]: There are too many decisions in the function about transactional logic. I think would be better to create separate function `install__transactional_kernel_debuginfo`. Added if-else decreased readability of logic in this case.

File: lib/kdump_utils.pm [hunk 4]
```diff
@@ -116,6 +132,21 @@ sub prepare_for_kdump {
         zypper_call("rr $snapshot_debuginfo_repo");
         return;
     }
+    if (is_alp) {
+        my $repo_url = 'https://updates.suse.com/SUSE/Products/ALP-Dolomite/1.0/x86_64/product_debug/';
```
Comment (czerw) [inline lib/kdump_utils.pm:136]: Hardcoded repositories is not optimal. I would use variable define at test suite level `REPO_TRANSACTIONAL_DEBUG' with ibs url.

File: lib/kdump_utils.pm [hunk 5]
```diff
@@ -116,6 +132,21 @@ sub prepare_for_kdump {
...
+        install_kernel_debuginfo;
+        zypper_call("rr debuginfo");
+        return;
+    }
+    if (is_sle_micro('=5.5')) {
+        my $repo_url = "http://download.suse.de/ibs/SUSE:/SLE-15-SP5:/Update:/Products:/Micro55/images/repo/SLE-Micro-5.5-x86_64-Media3/";
+        zypper_call("ar $repo_url debuginfo");
+        install_kernel_debuginfo;
+        zypper_call("rr debuginfo");
+        return;
```
Comment (czerw) [inline lib/kdump_utils.pm:147]: If you use variable, you make above logic smaller and not duplicate 

```
        zypper_call("ar $repo_url debuginfo");
        install_kernel_debuginfo;
        zypper_call("rr debuginfo");
```

**Comment by czerw** [comment]
We will need verification of all these jobs (just use clone refspec):
SLE 15-SP6 (yast): https://openqa.suse.de/tests/12306081
SLE 15-SP6(cli): https://openqa.suse.de/tests/12306314
JeOS: https://openqa.suse.de/tests/12120528
15-SP5(maintenance): https://openqa.suse.de/tests/12313469

File: lib/kdump_utils.pm [hunk 6]
```diff
@@ -92,15 +106,17 @@ sub prepare_for_kdump {
     $args{test_type} //= '';
 
     # disable packagekitd
-    quit_packagekit;
-    my @pkgs = qw(yast2-kdump kdump);
-    push @pkgs, qw(crash);
+    if (!is_transactional) {
```
Comment (czerw) [inline lib/kdump_utils.pm:109]: Maybe it deserves separate function `get_repo_url_for_kdump_transactiona`  as well.

File: lib/kdump_utils.pm (hunk 4 above)
Comment (Martchus) [reply lib/kdump_utils.pm:136]: This is the URL I was pointed at when asking for debug repos for ALP. I have no idea where I would find it on IBS. I found https://download.suse.de/ibs/SUSE:/ALP:/Products:/Dolomite:/1.0/standard/ but it doesn't seem to have the debug repo.

**Comment by Martchus** [comment]
I refactored the code a little bit to address some of your points. I also did verification runs with and without `REPO_TRANSACTIONAL_DEBUG` and it still works.

**Comment by Martchus** [comment]
Verification runs for scenarios mentioned in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17870#issuecomment-1737043849:

* https://openqa.suse.de/tests/12334109
* https://openqa.suse.de/tests/12334110
* https://openqa.suse.de/tests/12334111
* https://openqa.suse.de/tests/12334112

File: lib/kdump_utils.pm (hunk 4 above)
Comment (Martchus) [reply lib/kdump_utils.pm]: To answer my own question: Maybe this repository http://download.suse.de/ibs/SUSE:/ALP:/Source:/Standard:/Core:/1.0:/Build/standard/x86_64/ could be used for ALP. But I made it use `REPO_TRANSACTIONAL_DEBUG` as suggested anyways so this question will only be relevant once we add the test to the yaml schedule.

**Review by schlad — APPROVED** [review]

**Review by czerw — APPROVED** [review]

**Pull request merged**